File: Common/DB/Mongo/Component/MongoProxyComponent.cpp
//
// Created by 64658 on 2025/5/12.
//

#include "MongoProxyComponent.h"
#include "Node/Component/NodeComponent.h"
#include "Mongo/Service/MongoReadProxy.h"
#include "Mongo/Service/MongoWriteProxy.h"
#include "Cluster/Config/ClusterConfig.h"

namespace mongo_cmd
{
	constexpr char* MONGODB_INC = "MongoWriteProxy.Inc";
	constexpr char* MONGODB_INSERT = "MongoWriteProxy.Insert";
	constexpr char* MONGODB_UPDATE = "MongoWriteProxy.Update";
	constexpr char* MONGODB_DELETE = "MongoWriteProxy.Delete";
	constexpr char* MONGODB_SET_INDEX = "MongoWriteProxy.SetIndex";

	constexpr char* MONGODB_COUNT = "MongoWriteProxy.Count";
	constexpr char* MONGODB_FIND_ONE = "MongoWriteProxy.FindOne";

}

namespace acs
{
	MongoProxyComponent::MongoProxyComponent()
	{
		this->mNodeMgr = nullptr;
	}

	bool MongoProxyComponent::LateAwake()
	{
		std::string name1 = ComponentFactory::GetName<MongoReadProxy>();
		std::string name2 = ComponentFactory::GetName<MongoWriteProxy>();
		LOG_CHECK_RET_FALSE(this->mNodeMgr = this->GetComponent<NodeComponent>())
		LOG_CHECK_RET_FALSE(ClusterConfig::Inst()->GetServerName(name1, this->mReadName))
		LOG_CHECK_RET_FALSE(ClusterConfig::Inst()->GetServerName(name2, this->mWriteName))
		return true;
	}

	int MongoProxyComponent::Inc(const char* key, int value)
	{
		return this->Inc(nullptr, key, value);
	}

	int MongoProxyComponent::Inc(const char* tab, const char* key, int value)
	{
		int targetValue = 0;
		do
		{
			Node* mongoNode = this->mNodeMgr->Next(this->mWriteName);
			if (mongoNode == nullptr)
			{
				LOG_ERROR("not find service:{}", this->mWriteName);
				break;
			}
			json::w::Document request;
			if (tab != nullptr)
			{
				request.Add("tab", tab);
			}
			request.Add("key", key);
			request.Add("value", value);
			std::unique_ptr<json::r::Document> response = std::make_unique<json::r::Document>();
			if (mongoNode->Call(mongo_cmd::MONGODB_INC, request, response) != XCode::Ok)
			{
				break;
			}
			response->Get("value", targetValue);
		} while (false);
		return targetValue;
	}

	int MongoProxyComponent::UpdateOne(const char* tab, const json::w::Value& filter, const json::w::Value& updater)
	{
		Node* mongoNode = this->mNodeMgr->Next(this->mWriteName);
		if (mongoNode == nullptr)
		{
			LOG_ERROR("not find service:{}", this->mWriteName);
			return XCode::NotFoundActor;
		}
		json::w::Document request;
		{
			request.Add("tab", tab);
			request.Add("limit", 1);
			request.Add("filter", filter);
			request.Add("document", updater);
		}
		return mongoNode->Call(mongo_cmd::MONGODB_UPDATE, request);
	}

	int MongoProxyComponent::DeleteOne(const char* tab, const json::w::Value& filter)
	{
		Node* mongoNode = this->mNodeMgr->Next(this->mWriteName);
		if (mongoNode == nullptr)
		{
			LOG_ERROR("not find service:{}", this->mWriteName);
			return XCode::NotFoundActor;
		}
		json::w::Document request;
		{
			request.Add("tab", tab);
			request.Add("limit", 1);
			request.Add("filter", filter);
		}
		return mongoNode->Call(mongo_cmd::MONGODB_DELETE, request);
	}

	int MongoProxyComponent::Deletes(const char* tab, const json::w::Value& filter, int limit)
	{
		Node* mongoNode = this->mNodeMgr->Next(this->mWriteName);
		if (mongoNode == nullptr)
		{
			LOG_ERROR("not find service:{}", this->mWriteName);
			return XCode::NotFoundActor;
		}
		json::w::Document request;
		{
			request.Add("tab", tab);
			request.Add("limit", limit);
			request.Add("filter", filter);
		}
		return mongoNode->Call(mongo_cmd::MONGODB_DELETE, request);
	}

	int MongoProxyComponent::Insert(const char* tab, const json::w::Value& document)
	{
		Node* mongoNode = this->mNodeMgr->Next(this->mWriteName);
		if (mongoNode == nullptr)
		{
			return XCode::NotFoundActor;
		}
		json::w::Document request;
		{
			request.Add("tab", tab);
			request.AddArray("list")->Push(document);
		}
		return mongoNode->Call(mongo_cmd::MONGODB_INSERT, request);
	}

	int MongoProxyComponent::Insert(const char* tab, const pb::Message& document)
	{
		Node* mongoNode = this->mNodeMgr->Next(this->mWriteName);
		if (mongoNode == nullptr)
		{
			return XCode::NotFoundActor;
		}
		std::string result;
		if (!pb_json::MessageToJsonString(document, &result).ok())
		{
			return XCode::ProtoCastJsonFailure;
		}
		json::w::Document request;
		{
			request.Add("tab", tab);
			request.AddArray("list")->PushObject(result);
		}
		return mongoNode->Call(mongo_cmd::MONGODB_INSERT, request);
	}

	int MongoProxyComponent::SetIndex(const char* tab, const std::string& field, int sort, bool unique)
	{
		Node* mongoNode = this->mNodeMgr->Next(this->mWriteName);
		if (mongoNode == nullptr)
		{
			return XCode::NotFoundActor;
		}
		json::w::Document request;
		{
			request.Add("tab", tab);
			request.Add("key", field);
			request.Add("sort", sort);
			request.Add("unique", unique);
		}
		return mongoNode->Call(mongo_cmd::MONGODB_SET_INDEX, request);
	}

	int MongoProxyComponent::Count(const char* tab)
	{
		Node* mongoNode = this->mNodeMgr->Next(this->mReadName);
		if (mongoNode == nullptr)
		{
			return -1;
		}
		json::w::Document request;
		{
			request.Add("tab", tab);
		}
		std::unique_ptr<json::r::Document> response = std::make_unique<json::r::Document>();
		if (mongoNode->Call(mongo_cmd::MONGODB_COUNT, request, response) != XCode::Ok)
		{
			return -2;
		}
		return response->GetInt("count", 0);
	}

	int MongoProxyComponent::Count(const char* tab, const json::w::Value& filter)
	{
		Node* mongoNode = this->mNodeMgr->Next(this->mReadName);
		if (mongoNode == nullptr)
		{
			return -1;
		}
		json::w::Document request;
		{
			request.Add("tab", tab);
			request.Add("filter", filter);
		}
		std::unique_ptr<json::r::Document> response = std::make_unique<json::r::Document>();
		if (mongoNode->Call(mongo_cmd::MONGODB_COUNT, request, response) != XCode::Ok)
		{
			return -2;
		}
		return response->GetInt("count", 0);
	}

	int MongoProxyComponent::FindOne(const char* tab, const json::w::Value& filter, json::IObject* document)
	{
		std::unique_ptr<json::r::Document> response
				= std::make_unique<json::r::Document>();
		int code = this->FindOne(tab, filter, response);
		if (code != XCode::Ok)
		{
			return code;
		}
		if (!document->Decode(*response))
		{
			return XCode::ParseJsonFailure;
		}
		return XCode::Ok;
	}

	int MongoProxyComponent::FindOne(const char* tab, const json::w::Value& filter, std::unique_ptr<json::r::Document> & document)
	{
		Node* mongoNode = this->mNodeMgr->Next(this->mReadName);
		if (mongoNode == nullptr)
		{
			return XCode::NotFoundActor;
		}
		json::w::Document request;
		{
			request.Add("tab", tab);
			request.Add("filter", filter);
		}
		return mongoNode->Call(mongo_cmd::MONGODB_FIND_ONE, request, document);
	}

	int MongoProxyComponent::FindOne(const char* tab, const json::w::Value& filter,
			const std::vector<std::string>& fields, std::unique_ptr<json::r::Document> & document)
	{
		Node* mongoNode = this->mNodeMgr->Next(this->mReadName);
		if (mongoNode == nullptr)
		{
			return XCode::NotFoundActor;
		}
		json::w::Document request;
		{
			request.Add("tab", tab);
			request.Add("filter", filter);
			request.Add("fields", fields);
		}
		return mongoNode->Call(mongo_cmd::MONGODB_FIND_ONE, request, document);
	}
}

File: bin/script/Component/MongoProxyComponent.lua


local MongoProxyComponent = { }

local node = require("Node")
local MONGO_READ_PROXY = "MongoReadProxy"
local MONGO_WRITE_PROXY = "MongoWriteProxy"

---@param tab string
---@param documents table
---@return number
function MongoProxyComponent:InsertBatch(tab, documents)
    local id = node:Allot(MONGO_WRITE_PROXY)
    local code, response = node:Call(id, "MongoWriteProxy.Insert", {
        tab = tab,
        list = documents
    })
    return code ~= XCode.Ok and 0 or response.count
end

---@param tab string
---@param document table
---@return number
function MongoProxyComponent:InsertOne(tab, document)
    local id = node:Allot(MONGO_WRITE_PROXY)
    local code, response = node:Call(id, "MongoWriteProxy.Insert", {
        tab = tab,
        list = { document }
    })
    return code ~= XCode.Ok and 0 or response.count
end

---@param tab string
---@param filter table
---@param document table
---@param cmd string
---@return number
function MongoProxyComponent:UpdateOne(tab, filter, document, cmd)
    local id = node:Allot(MONGO_WRITE_PROXY)
    local code, response = node:Call(id, "MongoWriteProxy.Update", {
        tab = tab,
        cmd = cmd,
        multi = false,
        filter = filter,
        document = document
    })
    return code ~= XCode.Ok and 0 or response.count
end

---@param tab string
---@param filter table
---@return number
function MongoProxyComponent:DeleteOne(tab, filter)
    local id = node:Allot(MONGO_WRITE_PROXY)
    local code, response = node:Call(id, "MongoWriteProxy.Delete", {
          tab = tab,
          limit = 1,
          filter = filter
    })
    return code ~= XCode.Ok and 0 or response.count
end

---@param tab string
---@param documents table
---@return number
function MongoProxyComponent:Update(tab, documents)
    local id = node:Allot(MONGO_WRITE_PROXY)
    local code, response = node:Call(id, "MongoWriteProxy.Updates", {
        tab = tab,
        list = documents
    })
    return code ~= XCode.Ok and 0 or response.count
end

---@param tab string
---@param list table
---@return number
function MongoProxyComponent:Deletes(tab, list)
    local id = node:Allot(MONGO_WRITE_PROXY)
    local code, response = node:Call(id, "MongoWriteProxy.Deletes", {
        tab = tab,
        list = list
    })
    return code ~= XCode.Ok and 0 or response.count
end

---@param tab string
---@param field string
---@param sort number
---@param unique boolean
---@return boolean
function MongoProxyComponent:SetIndex(tab, field, sort, unique)
    local id = node:Allot(MONGO_WRITE_PROXY)
    local code = node:Call(id, "MongoWriteProxy.SetIndex", {
        tab = tab,
        key = field,
        sort = sort or 1,
        unique = unique or false
    })
    return code == XCode.Ok
end

---@param tab string
---@param field string
---@param value number
---@return number
function MongoProxyComponent:Inc(tab, filter, field, value, upsert)
    local id = node:Allot(MONGO_WRITE_PROXY)
    local code, response = node:Call(id, "MongoWriteProxy.Inc", {
        tab = tab,
        field = field,
        filter = filter,
        upsert = upsert or false,
        value = value or 1
    })
    return code ~= XCode.Ok and 0 or response.value
end

---@param tab string
---@param filter table
---@param document table
function MongoProxyComponent:Save(tab, filter, document, upsert)
    local id = node:Allot(MONGO_WRITE_PROXY)
    local code, response = node:Call(id, "MongoWriteProxy.Save", {
        tab = tab,
        filter = filter,
        document = document,
        upsert = upsert or false
    })
    return code ~= XCode.Ok and 0 or response.count
end

---@param tab string
---@param filter table
---@param fields table
---@return table | nil
function MongoProxyComponent:FindOne(tab, filter, fields)
    local id = node:Allot(MONGO_READ_PROXY)
    local code, response = node:Call(id, "MongoReadProxy.FindOne", {
        tab = tab,
        filter = filter,
        fields = fields
    })
    return code == XCode.Ok and response or nil
end

---@param tab string
---@param filter table
---@param fields table
---@param limit number
---@return table | nil
function MongoProxyComponent:Find(tab, filter, fields, limit)
    local id = node:Allot(MONGO_READ_PROXY)
    local code, response = node:Call(id, "MongoReadProxy.Find", {
        tab = tab,
        filter = filter,
        fields = fields,
        limit = limit,
    })
    if code ~= XCode.Ok then
        return nil
    end
    return response.list, response.cursor
end

function MongoProxyComponent:FindPage(tab, filter, fields, limit, page, sort)
    local id = node:Allot(MONGO_READ_PROXY)
    local code, response = node:Call(id, "MongoReadProxy.FindPage", {
        tab = tab,
        filter = filter,
        fields = fields,
        page = page or 1,
        limit = limit or 10,
        sort = sort
    })
    if code ~= XCode.Ok then
        return nil
    end
    return response.list
end

---@param tab string
---@param filter table
---@return number
function MongoProxyComponent:Count(tab, filter)
    local id = node:Allot(MONGO_READ_PROXY)
    local code, response = node:Call(id, "MongoReadProxy.Count", {
        tab = tab,
        filter = filter
    })
    if code ~= XCode.Ok then
        return 0
    end
    return response.count
end

---@param tab string
---@param cursor number
---@param count number
---@return table | nil
function MongoProxyComponent:GetMore(tab, cursor, count)
    local id = node:Allot(MONGO_READ_PROXY)
    local code, response = node:Call(id, "MongoReadProxy.GetMore", {
        tab = tab,
        cursor = cursor,
        batchSize = count
    })
    if code ~= XCode.Ok then
        return nil
    end
    return response.list
end

---@param tab string
---@param filter table
---@param count number
---@param fields table,
---@param batchSize number
---@return table | nil
function MongoProxyComponent:RandDocuments(tab, filter, count, fields, batchSize)
    local id = node:Allot(MONGO_READ_PROXY)
    local code, response = node:Call(id, "MongoReadProxy.Random", {
        tab = tab,
        filter = filter,
        count = count or 1,
        fields = fields,
        batchSize = batchSize
    })
    if code ~= XCode.Ok then
        return nil
    end
    return response.list
end

---@param tab string
---@param pipeline table
---@param batchSize number
---@return table | nil
function MongoProxyComponent:Aggregate(tab, pipeline, batchSize)
    local id = node:Allot(MONGO_READ_PROXY)
    local code, response = node:Call(id, "MongoReadProxy.Aggregate", {
        tab = tab,
        pipeline = pipeline,
        batchSize = batchSize
    })
    if code ~= XCode.Ok then
        return nil
    end
    return response.list
end

function MongoProxyComponent:ListDatabases()
    local id = node:Allot(MONGO_READ_PROXY)
    local code, response = node:Call(id, "MongoReadProxy.ListDatabases")
    print(code)
    if code ~= XCode.Ok then
        return nil
    end
    return response.list
end

function MongoProxyComponent:ListCollections(db)
    local id = node:Allot(MONGO_READ_PROXY)
    local code, response = node:Call(id, "MongoReadProxy.ListCollections", {
        db = db
    })
    if code ~= XCode.Ok then
        return nil
    end
    return response.list
end

---@param tab string
---@param filter table
---@param document table
---@param fields table
function MongoProxyComponent:FindModify(tab, filter, document, fields)
    local id = node:Allot(MONGO_WRITE_PROXY)
    local code, response = node:Call(id, "MongoWriteProxy.FindModify", {
        tab = tab,
        filter = filter,
        fields = fields,
        document = document
    })
    if code ~= XCode.Ok then
        return nil
    end
    return response
end

return MongoProxyComponent